Chapter3 study

Min-Su opened the pull request:
#1

Min-Su committed:
- Create Skeleton project
메이븐 프로젝트 생성 및 pom.xml springframwork 5.1.4 의존성 주입

Min-Su committed:
-프로젝트 초기화2

Min-Su committed:
- 챕터3 예제 3.1 ~ 3.9 작성

Min-Su committed:
WIPG

Min-Su committed:
-불필요한 파일 및 디렉토리 삭제

Min-Su committed:
- Spring을 이용한 DI
필요한 객체로 설정클래스 AppCtx 정의 및 스프링 컨테이너 생성
getBean() 메소드를 통해 빈 객체 사용

Min-Su committed:
-DI 주입 생성자 방식
객체의 생성자를 통해 의존 객체 주입

Min-Su committed:
-DI 주입 setter 방식
객체 생성시점이 아닌 setter 메소드로 의존 객체 주입

Min-Su committed:
-기본 데이터 타입 값 설정
기본 데이터 타입 값 설정하여 사용 하는 예제

Min-Su committed:
- ChangePasswordService, MemberRegisterService 테스트 코드 작성

monorisk commented on src/test/java/spring/ChangePasswordServiceTest.java:
```diff
@@ -0,0 +1,33 @@
+
+
+    @Test
+    public void testChangePassword() {
```
테스트 코드를 개발할 때 다음과 같은 포맷을 이용해봅시다
``` java
public void testTESTMETHODNAME_scenario() throw Exception {
  // given
  
  // when

  // then 
}
```

monorisk replied:
어떤 상황일 때

무엇을 하면

어떻게 된다

정도의 관점에서 테스트 코드 시나리오를 만드는게 도움이 됩니다.

나중에 이것을 읽어 보는 다른 개발자가, 테스트 대상의 동작을 이해 하는데도 도움이 됩니다.

훌륭한 테스트는 코드를 읽는 대상에게 설계 문서 같은 역할을 하게 해주는 것이라고 생각합니다.

monorisk commented on src/test/java/spring/ChangePasswordServiceTest.java:
```diff
@@ -0,0 +1,33 @@
+
+        changePwdSvc.changePassword(member.getEmail(), oldPsw, newPsw);
+
+        assertEquals(newPsw, member.getPassword());
```
`assertJ`를 잘 활용해봅시다.

monorisk commented on src/main/java/spring/VersionPrinter.java:
```diff
@@ -0,0 +1,19 @@
+        System.out.printf("이 프로그램의 버전은 %d.%d입니다.\n\n", majorVersion, minorVersion);
+    }
+
+    public void setMajorVersion(int majorVersion) {
```
보통 메서드의 선언 순서라고 하면

생성자
getter & setter
비지니스

정도가 됩니다.

monorisk reviewed (commented):
예외적인 테스트 시나리오를 더 만들어봅시다.

적어도 비지니스 내부에서 Exception을 던지는 것들이 있는데, 그것이 누락 되는 것은 불충분한 시나리오 일 것 같습니다.

또한 어떤 예외를 던지는 모듈을 사용하는 모듈을 테스트할 때, 하위 모듈에서 예외를 던졌을 때 어떤 행동을 해야 하는지에 대해서 정의 하는 것도 좋을 것 같습니다.

Min-Su committed:
- review 수정사항 적용

Min-Su committed:
-- coding convention 정리

Min-Su committed:
-- Chapter 4 의존자동주입
멤버, 메소드에 @Autowired 어노테이션 부여로 의존 자동주입

Min-Su committed:
-- Chapter 4 의존성자동주입
빈 생성 클래스에 @Autowired 적용
Config 클래스인 AppCtx 클래스 의존 주입 코드 수정

Min-Su committed:
-- Chapter 4 의존자동주입
같은 타입의 빈 객체가 여러개 있다면 스프링 컨테이너가 빈 생성시 에러 발생으로
예외처리된다.
이 때, 빈 객체와 주입 대상 객체에 @Qualifier 어노테이션 설정으로
주입될 의존 객체를 선택한다.

Min-Su committed:
-- Chapter4 의존성 자동주입
자동 주입과 명시적 의존 주입 간의 관계
수동으로 의존성을 주입해도 @Autowired 어노테이션이 우선됨

Min-Su committed:
-- Chapter4 의존자동주입
Chapter3, Chapter4 리뷰사항 수정

Min-Su committed:
-- Chapter5 컴포넌트 스캔
빈으로 등록될 클래스를 지정하는 @Component, @Component로 지정된 빈 클래스를 스캔하는 @ComponentScan
@ComponentScan 속성 값으로 basePackage = {"패키지경로"} 를 부여 할 수 있다.

Min-Su committed:
-- Chapter5 컴포넌트 스캔
@ComponentScan 어노테이션은 excludeFilters 속성을 주어 컴포넌트 스캔 대상을 제외하거나 포함할 수 있다
1) FilterType.REGEX - 정규표현식을 사용하여 제외 대상 지정
2) FilterType.ASPECTJ - AspectJ 패턴을 사용하여 제외 대상 지정
3) FilterType.ANNOTATION - 특정 어노테이션을 붙인 타입을 제외 대상 지정
4) FilterType.ASSIGNABLE_TYPE - 특정 타입이나 그 하위 타입을 제외대상 지정
*설정할 필터가 두 개 이상이면 속성에 배열 사용가능

monorisk commented on src/test/java/spring/MemberRegisterServiceTest.java:
```diff
@@ -16,15 +16,27 @@
 
+        //when
+
+        if (memberDao.selectByEmail("karosis28@gmail.com") != null) {
```
https://github.com/minsuhas/spring5/pull/2/files#diff-24fad7b5b39e6bce8ac22780c507be81R17 를 테스트 하기 위한 시나리오기 때문에

```java
@Test(expected = DuplicateMemberException.class)
public void testRegist_when_email_is_already_exist() throw Exception {
  // given
  memberRegSvc = ctx.getBean("memberRegSvc", MemberRegisterService.class);

  String alreadyExistEmail= "alreadyExistEmail@naver.com";

  RegisterRequest req = new RegisterRequest();
  req.setEmail(alreadyExistEmail);
  req.setPassword("1234");
  req.setName("kimminsu");

  // when
  memberRegSvc.regist(req);

  // then
  // throw DuplicateMemberException
}
```

monorisk committed:
Merge pull request #4 from minsuhas/chapter4

Chapter4 의존성자동주입
